lib/sport_ngin_aws_auditor/instance_helper.rb

```diff
@@ -109,18 +79,82 @@ def sort_through_RIs(client)
       return ris_region, ris_hash
     end
 
-    def measure_differences(instance_hash, ris_hash)
+    def measure_differences(instance_hash, ris_hash, ris_region, klass)
       differences = Hash.new()
+
+      # Setup and zone-based RI calculation
       instance_hash.keys.concat(ris_hash.keys).uniq.each do |key|
         instance_count = instance_hash.has_key?(key) ? instance_hash[key][:count] : 0
         ris_count = ris_hash.has_key?(key) ? ris_hash[key][:count] : 0
-        differences[key] = {:count => ris_count - instance_count, :region_based => false}
+        # positive count => more ris, negative count => more instances
+        differences[key] = { count: ris_count - instance_count, region_based: false }
       end
+
+      # Region-based RI calculation
+      ris_region_processor(differences, ris_region) unless ris_region.empty?
+
       differences
     end
 
+    # Assuming zone based information is already added and differences hash is initialized
+    private def ris_region_processor(differences, ris_region)
+      # Group all the same size RIs
+      # Ex: ri_group_arr = [{ instance_type: "t2.medium", platform: "Linux VPC", count: 7 }, ...]
+      ri_group_arr = group_ris_region(ris_region)
+
+      # Process each ri_group determined by instance_type and platform
+      ri_group_arr.each do |ri_group|
+        target_instance_type, target_platform, ri_count = ri_group[:instance_type], ri_group[:platform], ri_group[:count]
+        instances_left = differences.select do |k,v|
+          # Ex: k,v = "Linux VPC us-east-1d t2.small", {:count=>-15, :region_based=>false}
+          Regexp.new("^#{target_platform}.*#{target_instance_type}$") =~ k
+        end
+
+        instances_left_count = - instances_left.reduce(0) do | previous_count, actual_instance |
+          previous_count += actual_instance[1][:count]
+        end
+
+        if instances_left_count > 0
+          # Actual instances left > zone RIs
+          # Use regional instances
+          instances_left.each_value do |v|
+            if ri_count >= -v[:count]
+              ri_count += v[:count]
+              v[:count] = 0
+            else
+              # No more RIs left
+              v[:count] = ri_count + v[:count]
+              ri_count = 0
+              break;
+            end
+          end
+        end
+        differences["#{target_platform} #{target_instance_type}"] = { count: ri_count, region_based: true }
+      end
+    end
+
+    private def group_ris_region(ris_region)
+      ri_group_arr = []
+      ris_region.each do |ec2_ri_obj|
+        selected_ri_group = ri_group_arr.select { |ri_group| ri_group[:instance_type] == ec2_ri_obj.instance_type && ri_group[:platform] == ec2_ri_obj.platform  }
```

Comment:
On line 137 is when we instantiate `ri_group_arr`, so won't this array always come out with nothing in it?

Reply:
`ri_group_arr` is built as it processes each `ris_region`. So if `ris_region` is
```
[
#<SportNginAwsAuditor::EC2Instance:0x007fbfc9b339a0 @id="00a09b7d-3025-4fda-84b8-c174c43e4bb5", @name=nil, @platform="Linux VPC", @scope="Region", @availability_zone=nil, @instance_type="t2.small", @count=6, @stack_name=nil>,
#<SportNginAwsAuditor::EC2Instance:0x007fbfca4e5598 @id="0c92a282-8e51-4f36-906e-56b41f256788", @name=nil, @platform="Linux VPC", @scope="Region", @availability_zone=nil, @instance_type="t2.small", @count=4, @stack_name=nil>, 
#<SportNginAwsAuditor::EC2Instance:0x007fbfca444af8 @id="19e21c8d-24b3-4b71-8c67-0747f3cac98f", @name=nil, @platform="Linux VPC", @scope="Region", @availability_zone=nil, @instance_type="m5.large", @count=3, @stack_name=nil>, 
#<SportNginAwsAuditor::EC2Instance:0x007fbfcaa5ddb8 @id="1a23aad2-5fde-4892-a7d4-25e605fddf00", @name=nil, @platform="Linux VPC", @scope="Region", @availability_zone=nil, @instance_type="c5.large", @count=2, @stack_name=nil>
]
```
, `ri_group_arr` will be 
```
[
  { instance_type: "t2.small",
    platform: "Linux VPC",
    count: 10 },
  { instance_type: "m5.large",
    platform: "Linux VPC",
    count: 3 },
  { instance_type: "c5.large",
    platform: "Linux VPC",
    count: 2 }
]
```